File: User/ugcp_svc/Content/ProcessOutput/Commands/CNXT.cpp
#include "../../framework.h"
#include "CNXT.h"
#include "../../NXT/HIDNXTWrite.h"

/// <summary>
/// Gladiator NXT commands
/// </summary>
/// <returns><para>TRUE: processed and continue</para><para>FALSE: not processed</para></returns>
bool CNXT::Process(CEventPacket* queue)
{
	bool processed = true;
	PEV_COMMAND command = queue->GetCommandQueue()->front();

	if (command->Type.Get() == CommandType::NxtLeds)
	{
		UCHAR params[4]{};
		params[0] = command->Basic.Data;

		std::deque<PEV_COMMAND>::iterator pos = queue->GetCommandQueue()->begin();
		params[1] = (*(++pos))->Basic.Data;
		delete (*pos);
		queue->GetCommandQueue()->erase(pos);

		pos = queue->GetCommandQueue()->begin();
		params[2] = (*(++pos))->Basic.Data;
		delete (*pos);
		queue->GetCommandQueue()->erase(pos);

		pos = queue->GetCommandQueue()->begin();
		params[3] = (*(++pos))->Basic.Data;
		delete (*pos);
		queue->GetCommandQueue()->erase(pos);

		if (CNXTWrite::Get() != nullptr) CNXTWrite::Get()->SetLed(params);
	}
	else
	{
		processed = false;
	}

	return processed;
}

File: User/ugcp_svc/ProcessOutput/Commands/CNXT.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ugcp_svc.ProcessOutput.Commands
{
    static class CNXT
    {
        /// <summary>
        /// Gladiator NXT commands
        /// </summary>
        /// <returns><para>TRUE: processed and continue</para><para>FALSE: not processed</para></returns>
        public static bool Process(EventQueue.CEventPacket queue)
        {
            bool processed = true;

            ref EventQueue.EV_COMMAND command = ref System.Runtime.InteropServices.CollectionsMarshal.AsSpan(queue)[0];
            if (command.Type.Get() == EventQueue.CommandType.NxtLeds)
            {
                byte[] param = [0, 0, 0, 0];
                param[0] = command.Data.Basic.Data1;

                param[1] = queue[1].Data.Basic.Data1;
                queue.RemoveAt(1);

                param[2] = queue[1].Data.Basic.Data1;
                queue.RemoveAt(1);

                param[3] = queue[1].Data.Basic.Data1;
                queue.RemoveAt(1);

                NXT.CNXTWrite.Get()?.SetLed(param);
            }
            else
            {
                processed = false;
            }

            return processed;
        }
    }
}
